### ISSUE TITLE
Doctrine bridge

### ISSUE THREAD

**Issue by theofidry** [hautelook/AliceBundle#106]
I'm opening a new issue for this one as it seems to be a concern for you guys.

As mentioned in #102, there is a fundamental difference between AliceBundle and DoctrineFixturesBundle. For DoctrineFixturesBundle, a data loader is a fixture and you have to handle the persistence at the data loader level. In the 0.2 version of AliceBundle, the logic has been to add the possibility to include fixtures files if you wished for it and still rely on DoctrineFixturesBundle command.

For the 1.0, with the upgrade of nelmio/alice to the 2.0 which is a complete rewrite, some things changed. Relying on Doctrine command was just becoming a real pain and did not make much sense. The command is overly complex because of the data loaders resolutions, and to add new functionalities which were testable was very difficult.

But I can propose the following change: by default the `AbstractLoader` will load the fixtures specified by the `::getFixtures()`. It may be enough, but I'm certain there will be some edge cases that will not be handled well.

What do you think @Soullivaneuh, @guill3m and @yakobe?

/cc @h4cc

**Comment by Soullivaneuh** [comment]
> Relying on Doctrine command was just becoming a real pain and did not make much sense.

Could you elaborate?

**Comment by yakobe** [comment]
I'm not sure what you mean with
>But I can propose the following change: by default the AbstractLoader will load the fixtures specified by the ::getFixtures(). It may be enough, but I'm certain there will be some edge cases that will not be handled well.
What if the doctrine ::load method was just running some php to add some data?

Also, with this bundle command there seems to be no way of ordering the fixtures (or at least it says that the `Doctrine\Common\DataFixtures\OrderedFixtureInterfacez is not supported. Or did i miss something?

**Comment by theofidry** [comment]
> Could you elaborate?

@Soullivaneuh, if you take a look at the bundle Command, you'll see there is almost nothing in common. The Doctrine command has been designed for a specific need and extending it and make something testable is really hard. Plus the loading logic is completely different: the data loaders are loaded separately whereas for Alice this would cause issues for relations (see nelmio/alice#230), hence loading every files first then persist it.

> Also, with this bundle command there seems to be no way of ordering the fixtures (or at least it says that the Doctrine\Common\DataFixtures\OrderedFixtureInterface is not supported. Or did i miss something?

@yakobe, no there is no way to order fixtures because it no longer make sense. With Doctrine you have this because in you `::load()` method, you have the responsibility of persisting your fixtures, which might require some preexisting fixtures or depend on other one. So the order/dependences matters. With Alice, all the fixtures are loaded in one go so there is not this problem.

**Comment by theofidry** [comment]
> But I can propose the following change: by default the AbstractLoader will load the fixtures specified by the ::getFixtures(). It may be enough, but I'm certain there will be some edge cases that will not be handled well.
What if the doctrine ::load method was just running some php to add some data?

I mean something like:

```php
public function load(ObjectManager $objectManager)
{
    $this->container->get('loaderServiceId')->load($objectManager, $this->getFixtures());
}
```

But with nelmio/alice#230 issue, there is a guarantee to an issue at some point with relations.

Also Alice supports PHP fixtures files:

```php
<?php
// src/AppBundle/DataFixtures/ORM/user.php

return [
    'Nelmio\Alice\support\models\User' => [
        'user1' => [
            'username' => 'johnny',
            'favoriteNumber' => 42,
        ],
    ],
    'Nelmio\Alice\support\models\Group' => [
        'group1' => [
            'owner' => 1
        ],
    ],
];
```

**Comment by yakobe** [comment]
Ok so this command should replace the doctrine command. 
However, for the ordering: command will add all the fixtures in the order the bundles are defined. This can cause problems if certain yml files in one bundle need to be loaded before others. I could specify the bundles one by one to get the right order when running the command, but that is not practical every time.
Yes, we could change the order that the bundles are registered in AppKernal.php, but this seems a bit fragile?

**Comment by theofidry** [comment]
> Ok so this command should replace the doctrine command. 

Yes completely.

>  I could specify the bundles one by one to get the right order when running the command, but that is not practical every time.
Yes, we could change the order that the bundles are registered in AppKernal.php, but this seems a bit fragile?

It won't work. Even right now without changing anything, the order in which the files are loaded change from a computer to another, for instance from mine to Travis.

Do you have an example were you need to have control of the order?

**Comment by yakobe** [comment]
Lets say we have several bundles. One bundle is the core and loads users and other basic data.
Another bundle loads more data that references users. If these bundles are not loaded in the correct order then the yml files are loaded out of order and the reference cannot be found.

**Comment by theofidry** [comment]
@yakobe you're right I though Alice would handle that fine as I've never encountered the issue.

For the meantime the order specified in the `::getFixtures()` is kept, only the order in which the data loaders are retrieved is not.

But IMO it's not something the user should have to deal with. If a fixture file use a reference from another one, the loader should be able to resolve it.

**Comment by guill3m** [comment]
We have the same issue. You are right that the user should not have to deal with it. How is it possible to go about fixing this? Any ideas?

**Comment by theofidry** [comment]
I'll investigate this this week.

My idea for now (for what's it's worth): if I pass an array of fixtures files `$fixtures` and some file fails, an array of files which have failed to be loaded is build, then the loader retry to load them again and again until a certain limit is reached. This way if you have:

```
product.yml
brand.yml
dummy.yml
```

with `product.yml` which require `brand.yml`, it will fail the first time but work the second time.

That being said, chances are high that there is some cases pushing Alice beyond what it can currently do  some work on it will be needed. cc @Seldaek @tshelburne

**Comment by theofidry** [comment]
The problem of references across files will be solved with #115.
